Replaces Auxlua with the byondapi-based Dreamluau

Y0SH1M4S73R opened the pull request:
## About The Pull Request

Ever since byondapi went stable, I've been meaning to create a replacement lua library that uses it instead of the auxtools-based auxlua. After so many months, I've finally got the code just about into a position where it's ready for a PR.

[Click here](https://hackmd.io/@aloZJicNQrmfYgykhfFwAQ/BySAS18u0) for a guide to rewriting auxlua scripts for dreamluau syntax.

## Why It's Good For The Game

Code that runs on production servers should not depend on memory hacks that are liable to break any time Dream Daemon updates.

## Changelog

:cl:
admin: Admin lua scripting uses a new library that (probably) will not break when BYOND updates.
/:cl:

## TODO:
- [x] Convert the lua editor ui to TS
- [x] Include a guide for converting scripts from auxlua syntax to dreamluau syntax

Y0SH1M4S73R committed:
a bunch of stuff to switch auxlua out for dreamluau

Y0SH1M4S73R committed:
pre-rebase commit

Y0SH1M4S73R committed:
Update .gitignore

Y0SH1M4S73R committed:
Merge remote-tracking branch 'upstream/master' into new_lua_library

Y0SH1M4S73R committed:
another pre-rebase commit

Y0SH1M4S73R committed:
Merge remote-tracking branch 'upstream/master' into new_lua_library

Y0SH1M4S73R committed:
Update dreamluau.dll

Y0SH1M4S73R committed:
Merge remote-tracking branch 'upstream/master' into new_lua_library

Y0SH1M4S73R committed:
Update run_integration_tests.yml

SyncIt21 commented:
Fixes #80578

san7890 commented on dependencies.sh:
```diff
@@ -21,11 +21,11 @@ export SPACEMAN_DMM_VERSION=suite-1.8
 
-#auxlua repo
-export AUXLUA_REPO=tgstation/auxlua
+#dreamluau repo
+export DREAMLUAU_REPO="y0sh1m4s73r/dreamluau"
```
tell mothblocks or whomever headcoder to add this to the org liek we did for auxlua

Watermelon914 replied:
we already have one:
https://github.com/tgstation/auxlua

Watermelon914 replied:
then again I suppose it does have a different branding, we should definitely make a new one

SyncIt21 commented:
Don't let this go stale. @Y0SH1M4S73R

optimumtact commented:
This is actually mergeable as is right.

Y0SH1M4S73R commented:
I still have a couple little things to test and then commit before this is merge-ready.

Y0SH1M4S73R committed:
converts the lua editor ui to react typescript

Y0SH1M4S73R committed:
update the lua README.md

Y0SH1M4S73R committed:
fixes some small bugs on the DM side of lua code

SyncIt21 commented:
> Fixes #80578

It does right?

jlsnow301 commented:
Can this be TM'd?

Cyberboss commented:
Not easily on TGS3 i think

optimumtact commented:
yeah i seem to remember external dependencies being a pain to update.

optimumtact commented:
although this uses a new DLL so it should be fine.

Watermelon914 commented:
What's the lua execution limit with this change btw?

Watermelon914 commented on lua/SS13_base.lua:
```diff
@@ -67,123 +46,92 @@ function SS13.await(thing_to_call, proc_to_call, ...)
+	if not handler_data then
+		return
+	end
+	handler_data.functions[func] = nil
```
This causes a runtime if func is null.
The default behaviour (as set out before) should be that all signals on datum for signaltype `signal` should be unregistered. Otherwise, this is going to break the majority of my lua scripts

Watermelon914 commented:
Would be nice if the 'Load Script' option was added that allowed direct upload via byond.
My scripts have generally reached a size where it hits the topic byte limit and refuses to be sent to the server as a result.

Y0SH1M4S73R commented:
> Would be nice if the 'Load Script' option was added that allowed direct upload via byond. My scripts have generally reached a size where it hits the topic byte limit and refuses to be sent to the server as a result.
> 
> It existed before but seems to be removed in this update

It was in the "File" menu under "Run" but the act it called was incorrect. I'm fixing it and changing its name to be more clear

Y0SH1M4S73R committed:
fixes and clarifies script uploading

Y0SH1M4S73R committed:
`SS13.unregister_signal` removes all handlers when passed nil

Watermelon914 commented:
> It was in the "File" menu under "Run" but the act it called was incorrect. I'm fixing it and changing its name to be more clear

I see, thanks

Watermelon914 commented:
Save throws an error:
`TypeError: Object doesn't support this action`

Watermelon914 commented:
Also:
syntax error: lua/SS13_base.lua:138: Expected identifier when parsing expression, got '[' stack traceback:[C]: in ?[C]: in ?lua/SS13.lua:1: in ?[C]: in ?[string "input"]:1: in ?

Watermelon914 commented on lua/SS13_base.lua:
```diff
@@ -67,123 +46,107 @@ function SS13.await(thing_to_call, proc_to_call, ...)
-			clear_handler(handler_info)
+	if func == nil then
+		if signal == "parent_qdeleting" then
+			handler_data.functions = []
```
```suggestion
			handler_data.functions = {}
```

Watermelon914 commented on lua/timer.lua:
```diff
@@ -37,22 +37,21 @@ function __stop_internal_timer(func)
 			continue
 		end
-		if over_exec_usage(0.85) then
+		if _exec.time / _exec.limit > 0.85 then
 			sleep()
 		end
```
since this isn't such a low value as 100ms anymore, you may want to instead use tick_usage as a measure instead.
Otherwise, this'll probably end up eating all of the tick time.

Will probably need to look at implementing a library to handle performance since over_exec_usage isn't a super reliable indicator anymore to prevent the server from freezing up. It wasn't perfect before but now it definitely won't be any useful anymore.

Y0SH1M4S73R committed:
fix a syntax error in signal clearing

Y0SH1M4S73R committed:
rebases the timer execution limit on `world.tick_lag`

Y0SH1M4S73R committed:
removes the save button from the ui...

...because Dream Seeker prepends blob urls with `byond://` and then tries to connect to them as if they were byond worlds

Y0SH1M4S73R committed:
adds `blob-polyfill` as a tgui dependency...

...to add support for `Blob` methods like `text`

Y0SH1M4S73R committed:
i had to sort imports manually

Y0SH1M4S73R committed:
Merge branch 'master' into new_lua_library

Y0SH1M4S73R committed:
fixes calling global functions

Y0SH1M4S73R committed:
Update dependencies.sh

optimumtact commented:
thanks

Absolucy commented:
minimum version should be bumped to 515.1642 imo, due to [this bug](https://www.byond.com/forum/post/2936624)

Reveyni commented:
I'm trying to convert watermelon's auxlua codes to dreamlua, but I'm getting a lot of errors. Or most things don't work. Is it possible for you to share a more detailed document with more examples? I'm having a hard time understanding some things in the document you shared.
